content/blog/2025-12-16-facade-layout-compiler.md

```diff
@@ -0,0 +1,409 @@
+
+This classification prevents the compiler from being overly restrictive while still catching genuine architectural errors.
+
+---
+
+## Back-End: Code Generation
+
+For valid programs, the compiler generates two output formats:
+
+### JSON Output
+
+The JSON output includes:
+
+- Grid metadata (dimensions, rules, auto-filled cells)
+- The normalized grid as a 2D array of symbols
+- Validation results (is_valid, errors, warnings)
+
+This format is designed for programmatic consumption. The structured data can be used by downstream tools for further processing, constraint solving, or generative design.
+
+### SVG Output
+
+The SVG generator produces a visual representation where each cell is rendered as a colored rectangle with a label. Cells are grouped by type in the SVG DOM, making it easy to style or manipulate specific element types.
+
+### PNG Rasterization
+
+For environments that don't support SVG, the compiler can produce PNG output via CairoSVG.
+
+---
+
+## What were the hardest parts to get right?
+
+### 1. Separating Syntax Errors from Semantic Errors
+
+The distinction between lexical, syntactic, and semantic errors was initially blurry. For example:
+
+- Is `@` an invalid character (lexical) or an invalid symbol (semantic)?
+- Is `row 1: X E E` a syntax error or a semantic error?
+
+I resolved this by establishing boundaries:
+
+- **Lexical errors:** Characters not in the DSL's alphabet (punctuation, Unicode, emoji)
+- **Syntax errors:** Malformed constructs (missing colons, invalid repeat expressions, missing row numbers)
+- **Semantic errors:** Valid syntax with invalid meaning (floating doors, disconnected chimneys, invalid symbol letters like `X`)
+
+The key insight was that `X` is a valid identifier syntactically but not a valid symbol semantically. The parser accepts it and emits a semantic error during cell processing.
+
+### 2. Chimney Connectivity Analysis
+
+Initially, I tried a simple positional check (e.g., chimney cells must be on the top row). This failed for common patterns like diagonal chimneys:
+
+```
+row 1: C E E
+row 2: C C E
+row 3: E C C
+row 4: E E C
+```
+
+This chimney is valid. It connects to the roof through a diagonal path. The flood-fill approach correctly handles this by treating chimneys as connected components rather than isolated cells.
+
+A subtle bug emerged with multiple chimney components: the algorithm must check each component independently, not just verify that some chimney reaches the roof.
+
+### 3. Zero-Count Repeat Expressions
+
+The expression `E*0` presented an interesting design decision. Options:
+
+1. Treat as syntax error (reject)
+2. Treat as single cell (emit one `E`)
+3. Treat as zero cells (emit nothing)
+
+I chose option 3 because it's the most mathematically consistent interpretation and enables useful patterns like conditional exclusion. The implementation simply doesn't add any cell to the AST when count is 0.
```

Comment:
I think you've already covered this thoroughly above when you first introduced the feature.